DIRHASH_STANDARD.md

```diff
@@ -0,0 +1,203 @@
+4. **Implementation**: Additional, implementation specific, aspects of how the `DIRHASH` is programmatically obtained.
+
+Since the configuration options for 1-3 above affects the obtained `DIRHASH`, they must be represented in a [`DIRHASH` based checksum](#the-dirsum-object) of a directory. Subsequently, the naming and interpretation of these options must be standardized to allow for communication and verification of such checksums.
+It is not required that a given package/module, implementing the `DIRHASH` standard, strictly follows the option naming (to adhere to language-specific convention e.g.) as long as they are properly translated to the standard options.
+
+Implementation options are naturally not covered by the Dirhash Standard as these *should not affect the obtained* `DIRHASH`. Such options should be clearly distinguished from those in 1-3.
+
+
+
+## Hash Function
+All data and metadata which forms the basis for the `DIRHASH` is hashed by a standard hash function. The Dirhash Standard defines precisely how data is fed to the hash function. The hash function is specified using the single option `algorithm`. The following algorithms are covered in the Dirhash Standard: [`"md5"`](https://www.ietf.org/rfc/rfc1321.txt), [`"sha1"`](https://tools.ietf.org/html/rfc3174), [`"sha224"`, `"sha256"`, `"sha384"`, `"sha512"`](https://tools.ietf.org/html/rfc6234) but can naturally be extended.
+
+
+### Hash function Options
+Name  | Type             | Default             | Description 
+----  | ---------------- | ------------------- | -----------
+algorithm | String | (no default) | The standard hashing algorithm (function) to use. 
+
+
+
+
+## Filtering
+Filtering governs what files and subdirectories to include. This is done by matching of file paths as well as handling of symbolic links and empty directories.
+
+
+### Filtering Options
+Name  | Type             | Default             | Description 
+----  | ---------------- | ------------------- | -----------
+match_patterns | Array of Strings | `["*"]` (match all) | Wildcard matching. Path relative to the Dirhash root is matched against the provided patterns. The path must match at least one of the "match patterns" (*not* starting with `!`) and none of the "ignore patterns" (starting with `!`).
+linked_dirs | Boolean | `true` | Include (i.e. follow) symbolic links to directories.
+linked_files | Boolean | `true` | Include symbolic links to files.
+empty_dirs | Boolean | `false` | Include empty directories. A directory is considered empty if it contains no files or directories to include *given the Filtering Options*.
+
+
+
+
+## Protocol
+The `DIRHASH` of a directory is obtained by taking the underlying hash function's hexdigest of a `DIR-DESCRIPTOR` string. The `DIR-DESCRIPTOR` is composed by concatenation of an ordered sequence of `ENTRY-DESCRIPTOR`:s, each followed by a newline:
+
+```<entry-descriptor>\n<entry-descriptor>...<entry-descriptor>\n```,
```

Comment:
no trailing `\000` vs trailing \n here?

Reply:
Good point, that is a bit arbitrary. Will skip trailing AND use `\000\000` instead of `\n` I'm pretty sure it is possible to construct collisions otherwise! Won't elaborate on it now but it is anyway more clear that no filename can mimic the separator this way.

Reply:
sounds good, ship it! :ship:

Reply:
😂One year later...